src/main/java/org/reaktivity/nukleus/oauth/internal/OAuthRealms.java

```diff
@@ -194,58 +207,179 @@ public JsonWebKey lookupKey(
     {
         private static final int MAX_SCOPES = 48;
 
-        private final Map<String, Long> scopeBitsByName = new CopyOnWriteHashMap<>();
+        private final List<OAuthRealmInfo> realmInfos = new CopyOnWriteArrayList<>();
 
-        private final long realmId;
         private final String realmName;
 
-        private long nextScopeBit;
+        private int nextScopeBit = 0;
 
         private OAuthRealm(
             String realmName)
         {
             assert nextRealmBit < MAX_REALMS;
             this.realmName = realmName;
-            this.realmId = 1L << nextRealmBit++ << MAX_SCOPES;
         }
 
         private long resolve(
+            String audienceName,
+            String issuerName,
             String[] scopeNames)
         {
             long authorization = NO_AUTHORIZATION;
-            if(nextScopeBit + scopeNames.length < MAX_SCOPES)
+            if (nextScopeBit + scopeNames.length < MAX_SCOPES)
             {
-                authorization = realmId;
+                final OAuthRealmInfo realmInfo = computeInfoIfAbsent(issuerName, audienceName);
+                authorization = realmInfo.realmId;
                 for (int i = 0; i < scopeNames.length; i++)
                 {
-                    authorization |= scopeBitsByName.computeIfAbsent(scopeNames[i], this::assignScopeBit);
+                    authorization |= realmInfo.supplyScopeBit(scopeNames[i]);
                 }
             }
             return authorization;
         }
 
         private long lookup(
+            String issuerName,
+            String audienceName,
             String[] scopeNames)
         {
-            long authorization = realmId;
-            for (int i = 0; i < scopeNames.length; i++)
+            final OAuthRealmInfo info = getInfoByFilter(issuerName, audienceName);
+            long authorization = NO_AUTHORIZATION;
+            if(info != null)
             {
-                authorization |= scopeBitsByName.getOrDefault(scopeNames[i], 0L);
+                authorization = info.realmId;
+                for (int i = 0; i < scopeNames.length; i++)
+                {
+                    authorization |= info.getOrDefault(scopeNames[i], 0L);
```

Comment:
Default is always `0L` so move that logic inside `getOrDefault` method and rename `getOrDefault` to `long scopeBit(String)`.